src/Microsoft.Health.Fhir.Core/Features/Search/Converters/FhirTypedElementToSearchValueConverterManager.cs

```diff
@@ -36,5 +41,33 @@ public bool TryGetConverter(string fhirType, Type searchValueType, out ITypedEle
 
             return _converterDictionary.TryGetValue((fhirType, searchValueType), out converter);
         }
+
+        internal class ExtensionConverter<T> : ITypedElementToSearchValueConverter
+        {
+            private readonly IEnumerable<ITypedElementToSearchValueConverter> _underlying;
+            private readonly Type _type;
+
+            public ExtensionConverter(IEnumerable<ITypedElementToSearchValueConverter> underlying)
+            {
+                _underlying = underlying;
+                _type = underlying.First().SearchValueType;
+            }
+
+            public IReadOnlyList<string> FhirTypes => new List<string> { "Extension" };
+
+            public Type SearchValueType => _type;
+
+            public IEnumerable<ISearchValue> ConvertTo(ITypedElement value)
+            {
+                if (value == null)
+                {
+                    return new List<ISearchValue>();
+                }
+
+                var typed = value.Select("value").FirstOrDefault();
+                var converter = _underlying.Where(x => x.FhirTypes.Contains(typed.InstanceType)).First();
```

Comment:
>First() [](http://example.com/codeflow?start=97&length=7)

This has a chance to blow out.

Reply:
ExtensionConverter groups together multiple underlying converters.

There is chance what we trying to convert type A to type B, and we have converter for A to C.

We will have ExtensionConverter which takes (A) type (because we have A->C), but if somehow we looking to extract type B it will explode with null exception.

So it should be FirstOrDefault() and null check.

Reply:
Still working on how to best handle the case where we don't find the right converter. Ideally, we should throw an error during search parameter validation; however, right now, this will allow users to post the search parameter and reindexing will later fail.

Reply:
Or maybe that's not the case based on nature of relationship between Type and FhirType.

Reply:
A summary of my discussion with @Ivanidzo4ka:

If no matching converter is found, it means that the resource's extension type can't be converted to the type specified in the search parameter definition.

For example, no matching converter would be found if we posted a resource with the following extension:

```
{
    "resourceType": "Condition",
   ...
    "extension": [
        {
            "extension": [
                {
                    ...
                }
            ],
            "url": ...
        },
        {
            "extension": [
                {
                    "url": "serviceProviderId",
                    "valueId": "123456"
                },
                {
                    "url": "journeyId",
                    "valueId": "678910"
                },
            ],
            "url": "http://example.org/fhir/StructureDefinition/RobinsPRExample"
        }
    ],
...
}
```

and a search parameter with the following type:

```
{
   "resourceType":"SearchParameter",
   ...
   "base":[
      "Condition",
      "Patient",
      "Observation"
   ],
   "type":"string",
   "expression":"DomainResource.extension.where(url='http://example.org/fhir/StructureDefinition/RobinsPRExample').extension.where(url='journeyId')"
}
```

since we can't convert `FhirType` `Id` to `Type` `StringSearchValue`.

We can't throw an error during search parameter validation, since we do not know the extension type at this point. Another thing to consider: we could have multiple resources with different type-variations of this extension.

If we decided to throw an error in the `ExtensionConverter`, we would fail to reindex all resources if one or more resources had an extension with an incompatible type. We would also be unable to post any new resources with this extension.

For now, we have (tentatively) decided to simply return nothing if no compatible type is found. This will allow resources with the correct type to be properly indexed, and it won't block us from creating new resources with this extension. It will not, however, give information to the user as to why resources with an incompatible extension type aren't being picked up.

@Ivanidzo4ka is there anything I missed? @brendankowitz thoughts?

As Ivan recommended, I am planning to reach out to the FHIR community to see how others have handled this edge case. I'll post the link to the post once I create it.